src/lib/Libifl/pbsD_connect.c

```diff
@@ -470,12 +349,186 @@ __pbs_connect_extend(char *server, char *extend_data)
 	 * Disable Nagle's algorithm on the TCP connection to server.
 	 * Nagle's algorithm is hurting cmd-server communication.
 	 */
-	if (pbs_connection_set_nodelay(sock) == -1) {
-		closesocket(sock);
+	if (pbs_connection_set_nodelay(sd) == -1) {
+		CLOSESOCKET(sd);
 		pbs_errno = PBSE_SYSTEM;
 		return -1;
 	}
 
+	return sd;
+}
+
+/**
+ * @brief	Get the array of connections to all servers
+ *
+ * @param	void
+ *
+ * @return	void
+ * @retval	!NULL - success
+ * @retval	NULL - error
+ *
+ * @par Side Effects:
+ *	None
+ *
+ * @par MT-safe: Yes
+ */
+void *
+get_conn_servers(void)
+{
+	svr_conn_t *conn_arr = NULL;
+
+	conn_arr = pthread_getspecific(psi_key);
+	if (conn_arr == NULL && pbs_conf.psi != NULL) {
+		int num_svrs;
+		int i;
+
+		num_svrs = get_num_servers();
+		conn_arr = calloc(num_svrs, sizeof(svr_conn_t));
```

Comment:
It would be better to store pointers to the structure of type svr_conn_t i.e. basically making conn_arr as follows
svr_conn_t **conn_array = NULL;

If we do this we can make use of the following utility functions.

- void * add_ptr_to_array(void *ptr_arr, void *ptr) 
- int remove_ptr_from_array(char **arr, char *str)
- void ** dup_array(void *ptr)
- int count_array(void *arr)

We need most of these functions for the following scenarios.

- When a Server is decommissioned i.e. When one of the servers is removed. We need to shrink the array for which we can use  remove_ptr_from_array
- When a new server is added. We to need to realloc the array and add details of new servers for which we can use add_ptr_to_array

Scheduler code already makes use of these functions wherever needed. These are actually part of Scheduler/misc.c
In fact we can make this part of Libutil.

Reply:
> When a Server is decommissioned i.e. When one of the servers is removed. We need to shrink the array for which we can use remove_ptr_from_array
When a new server is added. We to need to realloc the array and add details of new servers for which we can use add_ptr_to_array

When an admin changes the PBS_SERVER_INSTANCES list to decommission/add a new server, we would want to recreate the server connections array instead of modifying the existing array, right? Users might change the order of servers in the PBS_SERVER_INSTANCES list, so better to reallocate. This will be rare and the list will be small, so there'll be minimum performance hit. So, I think we are good there. Once we move to C++, we can use a vector for this list if needed. Please let me know what you think.

Reply:
Now that we are thinking to re-use the same slot for a server which gone down and came back and also going to use same connection array for scheduler I think we are good here. Also in the first phase if somebody wants to decommission a server, we are asking Admin to restart scheduler this is because he can inadvertently change the order of servers in PBS_SERVER_INSTANCES we are good here. May be in future based on the need we can revisit this and use a vector like array if needed.